src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -801,17 +803,38 @@ public override int GetHashCode()
 
         public bool Equals(Guid g) => EqualsCore(this, g);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static bool EqualsCore(in Guid left, in Guid right)
         {
-            ref int rA = ref Unsafe.AsRef(in left._a);
-            ref int rB = ref Unsafe.AsRef(in right._a);
+            if (Sse2.IsSupported)
+            {
+                var g1 = Unsafe.As<Guid, Vector128<byte>>(ref Unsafe.AsRef(left));
+                var g2 = Unsafe.As<Guid, Vector128<byte>>(ref Unsafe.AsRef(right));
+
+                if (Sse41.IsSupported)
+                {
+                    var xor = Sse2.Xor(g1, g2);
+                    return Sse41.TestZ(xor, xor);
+                }
+
+                var result = Sse2.CompareEqual(g1, g2);
+                return Sse2.MoveMask(result) == 0b1111_1111_1111_1111;
```

Comment:
```suggestion
                return g1.Equals(g2);
```
Results in same codegen and is simpler to read.
If this hinders inlining, etc. keep it as is (except `var` -> concrete type).

Ideally the JIT should emit code that takes SSE4.1's `_mm_testz_si128` into account?
(Iif this variant is always faster on the supported cpus)

Reply:
`Vector128<T>.Equals(Vector128<T>)` currently does not have a specific/optimized code path for SSE 4.1's `_mm_testz_si128`, which I why I called out an explicit code path for SSE 4.1 in `Guid.EqualsCore`.

But yes, under the SSE2 code path, it should be equivalent to invoke `g1.Equals(g2)` on the assumption that the JIT will inline the call, which it should because `Vector128<T>.Equals(Vector128<T>)`  is decorated with "aggressive inlining".

Reply:
> currently does not have a specific/optimized code path for SSE 4.1's

Yeah, my question was intented for @tannergooding (?) to have a look at this on the JIT-side 😃

Reply:
> Ideally the JIT should emit code that takes SSE4.1's _mm_testz_si128 into account?

There are a few improvements that could happen for `Vector128.Equals`. I've been waiting on https://github.com/dotnet/runtime/issues/49397 before tackling any of them, however.